dia/ideas2.rkt

```diff
@@ -0,0 +1,57 @@
+#lang racket
+
+(provide (rename-out [mb #%module-begin]) #%datum #%top #%top-interaction #%app)
+
+#| (Unstable) Module language for ideas
+
+Format:
+  #lang abe/ideas2 exported-ideas-id
+
+  - Ideas appraisal tree [N%]
+
+Meaning:
+
+- The body is an "appraisal" tree. Roughly, markdown bulleted list in tree
+form, with bracketed percentags [N%] at the end of the line.
+- Validate appraisals.
+- Tally idea attributions.
+- Validate attributions.
+- Export (provide) the exported-ideas-id bound to the attributions.
+
+|#
+
+(require syntax/parse/define
+         abe/dia)
+
+(define-syntax-parse-rule
+  (mb export:id {~datum <=} tree-data:expr)
+  (#%module-begin
+   (provide export)
+   (define tree 'tree-data)
+   (unless (validate-appraisal tree)
+     (error 'validate-appraisal "bad appraisal: ~a" tree))
+   (define export (make-hash))
+   (tally tree node-weight bump #:results export)
+   (unless (validate-attributions export)
+     (error 'validate-attributions "bad attributions: ~a" export))))
+
+(module reader syntax/module-reader abe/ideas2
+  #:whole-body-readers? #t
+  #:read reader
+  #:read-syntax syntax-reader
```

Comment:
The module language here is `syntax/module-reader`: https://docs.racket-lang.org/syntax/reader-helpers.html#%28mod-path._syntax%2Fmodule-reader%29

Reply:
Specifically, `#:whole-body-readers? #t` says that `reader` and `syntax-reader` are called _once_ to read the whole body, rather than repeatedly à la `read`/`read-syntax` from `racket/base`.